metapool/prep.py

```diff
@@ -0,0 +1,424 @@
+                      'characters: %s' %
+                      ', '.join(['"%s"' % i for i in invalid.values]))
+
+
+def preparations_for_run(run_path, sheet, pipeline=None):
+    """Given a run's path and sample sheet generates preparation files
+
+    Parameters
+    ----------
+    run_path: str
+        Path to the run folder
+    sheet: sample_sheet.SampleSheet
+        Sample sheet to convert
+    pipeline: str, optional
+        Which pipeline generated the data. The important difference is that
+        `atropos-and-bowtie2` saves intermediate files, whereas
+        `fastp-and-minimap2` doesn't. Defaults to `fastp-and-minimap2`, the
+        latest version of the sequence processing pipeline.
+
+    Returns
+    -------
+    dict
+        Filename to preparation file dictionary. Preparation files are
+        represented as DataFrames.
+    """
+
+    if pipeline is None:
+        pipeline = 'fastp-and-minimap2'
+
+    _, run_id = os.path.split(os.path.normpath(run_path))
+    run_date, instrument_code = parse_illumina_run_id(run_id)
+    instrument_model, run_center = get_model_and_center(instrument_code)
+
+    output = {}
+
+    required_columns = {'well_description', 'sample_plate',
+                        'sample_well', 'i7_index_id', 'index',
+                        'i5_index_id', 'index2'}
+
+    not_present = required_columns - set(sheet.columns)
+    if not_present:
+        warnings.warn('These required columns were not found %s'
+                      % ', '.join(not_present), UserWarning)
+        for col in not_present:
+            if col == 'well_description':
+                warnings.warn("Using 'description' instead of "
+                              "'well_description' because that column "
+                              "isn't present", UserWarning)
+                sheet[col] = sheet['description'].copy()
+            else:
+                sheet[col] = 'MISSING_FROM_THE_SAMPLE_SHEET'
+
+    for project, project_sheet in sheet.groupby('sample_project'):
+
+        project_name = remove_qiita_id(project)
+        qiita_id = project.replace(project_name + '_', '')
+
+        # if the Qiita ID is not found then make for an easy find/replace
+        if qiita_id == project:
+            qiita_id = 'QIITA-ID'
+
+        for lane, lane_sheet in project_sheet.groupby('lane'):
+
+            # this is the portion of the loop that creates the prep
+            data = []
+            for sample_name, sample in lane_sheet.iterrows():
+                run_prefix = get_run_prefix(run_path, project, sample_name,
+                                            lane, pipeline)
+
+                # we don't care about the sample if there's no file
+                if run_prefix is None:
+                    continue
+
+                row = {c: '' for c in PREP_COLUMNS}
+
+                row["sample_name"] = sample.well_description
+
+                row["experiment_design_description"] = "EXPERIMENT_DESC"
```

Comment:
Should the constants here, like EXPERIMENT_DESC or LIBRARY_PROTOCOL be defined as globals?

Reply:
Good call.